### PR TITLE
Site ID vs Coordinate conversions

### LINKED ISSUE #33: Implement to_index/to_indices for Coordinate type
Would allow users to index an array using the Coordinate type

### PR DESCRIPTION
**Pull request by emerali**
Resolves #33 (kinda)

### CODE REVIEW

File: src/types.jl [hunk 1]
```diff
@@ -13,13 +11,36 @@ For a more concrete definition please refer the following material:
...
+function size(::Val{true}, l::AbstractLattice, dim::Int)
+    if dim > ncoordinates(l)
+        return 1  # copying behaviour of `size` for multidimensional arrays
+    elseif dim == ncoordinates(l)
+        return unitcellsize(l)
+    else
+        return l.dims[dim]
+    end
+end
+size(::Val{false}, l::AbstractLattice, dim::Int) = (dim > ndims(l)) ? 1 : l.dims[dim]
```
Comment (Roger-luo) [inline src/types.jl:30]: could you remind me what is the first argument for?

File: src/sites.jl [hunk 2]
```diff
@@ -1,11 +1,39 @@
...
+
+checkbounds(::Type{Bool}, L::AbstractLattice, I::Int...) =
+    all((s, i)::NTuple{2,Int} -> checkindex(Bool, 1:s, i), zip(size(L), I))
+
+checkbounds(L::AbstractLattice, I::Int...) = checkbounds(Bool, L, I...) || throw(BoundsError(L, I))
+
+
+@inline function to_site_id(lattice::AbstractLattice, coords::NTuple{N}) where N
+    @boundscheck checkbounds(lattice, coords...)
+    (N == ncoordinates(lattice)) || ArgumentError("Incorrect number of coordinates! Lattice expects $(ncoordinates(lattice)), got $N")
```
Comment (Roger-luo) [inline src/sites.jl:11]: mark this as `@boundscheck` too?

**Review by Roger-luo — COMMENTED** [review]
the `Val{false}` thing looks a bit strange to me, might be better to just define an empty struct as holy trait.

File: src/types.jl (hunk 1 above)
Comment (emerali) [reply src/types.jl:30]: whether or not the lattice has a non-trivial unitcell

**Pull request merged**